tests/testPsfCandidateSelection.py

```diff
@@ -0,0 +1,120 @@
+        testDir = os.path.dirname(__file__)
+        self.exposure = afwImage.ExposureF(os.path.join(testDir, "data", "v695833-e0-c000-a00.sci.fits"))
+        Log.getLogger("characterizeImage").setLevel(Log.ERROR)
+
+    def tearDown(self):
+        del self.exposure
+
+    def testFlags(self):
+        """Simple test that the flags are being set."""
+        task = CharacterizeImageTask()
+        task.measurePsf.config.reserveFraction = 0
+        #   test that by default, no stars are reserved and used < candidates
+        results = task.characterize(self.exposure)
+
+        candidates = 0
+        used = 0
+        reserved = 0
+        for source in results.sourceCat:
+            if source.get("calib_psfCandidate"):
+                candidates += 1
+            if source.get("calib_psfUsed"):
+                used += 1
+            if source.get("calib_psfReserved"):
+                reserved += 1
+        self.assertLessEqual(used, candidates)
+        self.assertEqual(reserved, 0)
+
+    def testReserved(self):
+        """Test that a fraction can be reserved, and that the setting of
+           the exposureIdInfo will affect the set of candidates chosen."""
+        task = CharacterizeImageTask()
+        reservedSum = 0
+        #   set the reserve fraction, and see if the right proportion are reserved.
+        task.measurePsf.config.reserveFraction = .3
+        exposureIdInfo = ExposureIdInfo(12345, expBits=18)
+        results = task.characterize(self.exposure, exposureIdInfo=exposureIdInfo)
+        candidates = 0
+        reserved = 0
+        used = 0
+        for source in results.sourceCat:
+            if source.get("calib_psfCandidate"):
+                candidates += 1
+            if source.get("calib_psfUsed"):
+                used += 1
+            if source.get("calib_psfReserved"):
+                reservedSum += source.getId()
+                reserved += 1
+        self.assertEqual(reserved, int(.3 * (candidates + reserved)))
+        self.assertLessEqual(used, (candidates - reserved))
+
+        #   try again with a different id, and see if the list is different
+        task.measurePsf.config.reserveFraction = .3
+        exposureIdInfo = ExposureIdInfo(67890, expBits=18)
+        results = task.characterize(self.exposure, exposureIdInfo=exposureIdInfo)
+        candidates = 0
+        reserved = 0
+        used = 0
+        for source in results.sourceCat:
+            if source.get("calib_psfCandidate"):
+                candidates += 1
+            if source.get("calib_psfUsed"):
+                used += 1
+            if source.get("calib_psfReserved"):
+                reservedSum -= source.getId()
+                reserved += 1
+        self.assertEqual(reserved, int(.3 * (candidates + reserved)))
+        self.assertLessEqual(used, (candidates - reserved))
+        #   this sum will not generally be 0 if the set of reserved candidates is different
+        self.assertNotEqual(reservedSum, 0)
```

Comment:
I will change this, though I don't think that it can fail accidently unless the underlying catalog is changed.